### DISCUSSION TITLE
Using scripts to process data from the FlightGear simulator

### DISCUSSION THREAD

**Discussion by brightproject** [Mayitzin/ahrs#121] [Q&A]
I receive acceleration and angular velocity data from the `FlightGear` [simulator](https://github.com/JSBSim-Team/jsbsim/discussions/1084) via serial port.
Firstly, I couldn't get the `WSL` scripts to work in `Windows`.
Python3 simply didn't want to run the setup as described here.

> https://github.com/Mayitzin/ahrs?tab=readme-ov-file#installation

```
 python -m pip install .
Command 'python' not found, did you mean:
  command 'python3' from deb python3
  command 'python' from deb python-is-python3
```
That's why I'm just trying to run the `Python` code via `PyCharm`, but I haven't found a single example of working with the `AHRS` library in the folders themselves.
How can I insert data into it to determine orientation?
For example, I insert acceleration and angular velocity into the `Mahony` filter, either from the `MEMS` sensor or from the *csv file.
I don't know how to use the `AHRS` code in `Python`, although I have sufficient experience programming in `C++`.
This is not a complaint, it's just that the author of the library didn't really describe working with the library, only general phrases.
Also, the very first [issue](https://github.com/Mayitzin/ahrs/discussions/13) on GitHub says that the author will help everyone who is interested in the `AHRS` library, but in fact, most of the questions here are unanswered, which suggests that the project has been abandoned.
Please change me for such a tone, but until I can figure out this code, it will remain negative (tone).
Sorry!

**Comment by brightproject** [comment]
When I started writing this discussion, I knew that sooner or later I would figure out a little bit about the algorithms.
I found a little more information in another repository

> https://github.com/FlyTheThings/ahrs-1

I managed to run some examples right in the console.
As far as I understand, this way you can get a rotating matrix of cosines:
![Скриншот 07-09-2024 16 39 24](https://github.com/user-attachments/assets/b04c006c-ce27-4134-a064-3d5c74b2db94)

**Comment by brightproject** [comment]
I changed the example code a bit `new_test_filters.py`
But nothing is displayed on the graphs, actually I asked about such examples in this discussion... it seems like there is a lot of information, but at the same time there is nothing specific about the project.

```
import numpy as np
import ahrs
import matplotlib.pyplot as plt

RAD2DEG = ahrs.common.RAD2DEG
DEG2RAD = ahrs.common.DEG2RAD

class Test_Filter:
    def __init__(self, **kwargs):
        self.file = 'repoIMU.csv'
        self.data = ahrs.utils.io.load(self.file)


    def check_integrity(self, Q):
        sz = Q.shape
        qts_ok = not np.allclose(np.sum(Q, axis=0), sz[0]*np.array([1., 0., 0., 0.]))
        qnm_ok = np.allclose(np.linalg.norm(Q, axis=1).mean(), 1.0)
        return qts_ok and qnm_ok

    def allocate_arrays(self):
        self.Q = np.zeros((self.data.num_samples, 4))
        self.Q[:, 0] = 1.0

    def plot_quaternions(self, title):
        time = np.arange(self.data.num_samples)
        plt.figure(figsize=(12, 8))
        plt.subplot(2, 2, 1)
        plt.plot(time, self.Q[:, 0], label='q0')
        plt.title(title + " - q0")
        plt.xlabel('Time')
        plt.ylabel('q0')
        plt.grid()

        plt.subplot(2, 2, 2)
        plt.plot(time, self.Q[:, 1], label='q1', color='orange')
        plt.title(title + " - q1")
        plt.xlabel('Time')
        plt.ylabel('q1')
        plt.grid()

        plt.subplot(2, 2, 3)
        plt.plot(time, self.Q[:, 2], label='q2', color='green')
        plt.title(title + " - q2")
        plt.xlabel('Time')
        plt.ylabel('q2')
        plt.grid()

        plt.subplot(2, 2, 4)
        plt.plot(time, self.Q[:, 3], label='q3', color='red')
        plt.title(title + " - q3")
        plt.xlabel('Time')
        plt.ylabel('q3')
        plt.grid()

        plt.tight_layout()
        plt.show()

    def fourati(self):
        self.allocate_arrays()
        fourati = ahrs.filters.Fourati()
        for t in range(1, self.data.num_samples):
            self.Q[t] = fourati.update(DEG2RAD*self.data.gyr[t], self.data.acc[t], self.data.mag[t], self.Q[t-1])
        if self.check_integrity(self.Q):
            self.plot_quaternions("Fourati Filter")

    def ekf(self):
        self.allocate_arrays()
        ekf = ahrs.filters.EKF()
        for t in range(1, self.data.num_samples):
            self.Q[t] = ekf.update(DEG2RAD*self.data.gyr[t], self.data.acc[t], self.data.mag[t], self.Q[t-1])
        if self.check_integrity(self.Q):
            self.plot_quaternions("EKF Filter")

    def mahony(self):
        self.allocate_arrays()
        mahony = ahrs.filters.Mahony()
        for t in range(1, self.data.num_samples):
            self.Q[t] = mahony.updateMARG(DEG2RAD*self.data.gyr[t], self.data.acc[t], self.data.mag[t], self.Q[t-1])
        if self.check_integrity(self.Q):
            self.plot_quaternions("Mahony Filter")

    def madgwick(self):
        self.allocate_arrays()
        madgwick = ahrs.filters.Madgwick()
        for t in range(1, self.data.num_samples):
            self.Q[t] = madgwick.updateMARG(DEG2RAD*self.data.gyr[t], self.data.acc[t], self.data.mag[t], self.Q[t-1])
        if self.check_integrity(self.Q):
            self.plot_quaternions("Madgwick Filter")

```

I also tried to simply read data from a file using this code:

```
import numpy as np

# Load data from file
data = np.genfromtxt('repoIMU.csv', delimiter=';', skip_header=2)

# Extract time and quaternions
time = data[:, 0]
quaternions = data[:, 1:5]

# Extract accelerometer, gyroscope, and magnetometer data
acceleration = data[:, 5:8]
gyroscope = data[:, 8:11]
magnetometer = data[:, 11:14]

# Print data for verification
print("Time:", time)
print("Quaternions:", quaternions)
print("Acceleration:", acceleration)
print("Gyroscope:", gyroscope)
print("Magnetometer:", magnetometer)
```
The result in the console, after command `python3 new_test_filters.py` is as follows:

> Time: [8.000e-02 9.000e-02 1.000e-01 ... 8.998e+01 8.999e+01 9.000e+01]
> Quaternions: [[ 1.0000e+00  0.0000e+00  1.6232e-35  0.0000e+00]
>  [ 1.0000e+00  1.4021e-06  1.8673e-06 -2.9199e-07]
>  [ 1.0000e+00  2.8249e-06  3.7031e-06 -2.5293e-07]
>  ...
>  [ 1.6123e-01  9.8625e-01 -3.6266e-02 -1.3439e-03]
>  [ 1.5934e-01  9.8656e-01 -3.6239e-02 -1.3452e-03]
>  [ 1.5747e-01  9.8686e-01 -3.6211e-02 -1.3442e-03]]
> Acceleration: [[-0.071594  0.21157   9.7958  ]
>  [-0.071702  0.21878   9.7911  ]
>  [-0.081245  0.20916   9.7764  ]
>  ...
>  [-0.023582  2.559    -9.1003  ]
>  [-0.051928  2.5768   -8.879   ]
>  [-0.07637   2.61     -8.8695  ]]
> Gyroscope: [[ 2.3140e-03 -6.3400e-03  1.3220e-03]
>  [ 5.3300e-04 -7.2870e-03 -9.5700e-04]
>  [ 9.3420e-03 -4.2400e-04  3.5850e-03]
>  ...
>  [ 3.6318e-01 -1.2522e-02 -3.7300e-04]
>  [ 3.7517e-01 -8.7670e-03 -1.5938e-02]
>  [ 3.7059e-01 -2.6818e-02 -1.7078e-02]]
> Magnetometer: [[-0.33533  0.19856 -0.88708]
>  [-0.33402  0.20025 -0.88472]
>  [-0.33514  0.19707 -0.88613]
>  ...
>  [-0.35902 -0.45602  0.83672]
>  [-0.35677 -0.45603  0.83675]
>  [-0.35742 -0.4499   0.84045]]

I'll try, there are no really working examples to understand this zoo of functions.
For a newbie in `Python` and in general in the topic of `AHRS`, this is very, very bad.

**Comment by Mayitzin** [comment]
Hi! Sorry about the lack of clarity, especially for beginners, in the repository. You have raised a good point, and I think I can offer a good solution for it.

I will add a Jupyter Notebook with the most common uses of the functions and tools included in the repository. This way I can explain the structure of the package and its possible applications.

I hope I can improve the readability for everyone, and its use at all levels.

I'll write back here to ask for feedback about the notebook.

Mario.

**Reply by brightproject** [reply]
> I'll write back here to ask for feedback about the notebook.

@Mayitzin I just like to be indignant, that's my nature.
Perhaps, in the process of sorting out the code, I will be able to write several examples of working with the library.
Ideally, I want to get a solution to place the data received from the `flight simulator` into the code algorithms and receive data in the form of graphs or in the form of orientation angles, which can be written to a file or output to display devices.
The library of algorithms that you have collected is truly colossal, but as you correctly understood the problem - it is necessary to make a convenient `entry point` for using this library.
I am sure that professional programmers write similar algorithms themselves, and it is precisely to popularize your work and these algorithms that ready-made examples of an applied nature need to be made.
Thank you for your keen interest and participation in the discussion!

**Comment by Mayitzin** [comment]
Hi!

I added a Jupyter Notebook (https://github.com/Mayitzin/ahrs/blob/master/notebooks/Showcase.ipynb), where I will be showcasing the most common tools of the library, and I will try to make it as accessible as possible. If you think there should be something to further clarify, please let me know so that I can add it.

**Reply by brightproject** [reply]
> I added a Jupyter Notebook

This is great news, thank you for your work.

> If you think there should be something to further clarify, please let me know so that I can add it.

I was thinking of adding the ability to transfer raw data from accelerometers and gyroscopes from a flight simulator `FlightGear` to the algorithm `AHRS` on a `PC` and for the algorithm to process these values ​​in `Python` and output orientation angles to the serial port.
A kind of `SITL`(Software in the Loop) system.
Now I am processing data from the simulator on a real/physical microcontroller in `C++`, using the `RTIMULib2` algorithm.

https://github.com/user-attachments/assets/ff6a2e64-0263-4023-849a-7f4414c7cae7

It would be interesting to run the flight simulator, enable processing in Python using the `AHRS` library and get orientation angles and configure the filter, add a magnetic field through the `WMM`(World Magnetic Model), add noise, like real `MEMS` sensors.
In the code of the `AHRS` algorithm there is a `WMM`, as far as I remember, it is possible to use it and obtain a simulation of the magnetic field strength using the rotation matrices of a known magnetic field based on the `WMM` model and knowing the `Euler` rotation angles.
It seems to me that such a package would be interesting.
That is, real examples that can be used in practice.
I have a data parser and how to configure the output from the simulator `FlightGear`, I can send you the files and tell you how to integrate it.
I do not yet have an understanding of how to feed:

> Ax, Ay, Az, Gx, Gy, Gz

data to the `AHRS` code in `Python`, you understand and sort this out better than me🙂
